dashboard/config/initializers/mysql_avoid_utf8_alias_in_schema_dump.rb

```diff
@@ -0,0 +1,49 @@
+# In MySQL 5.7 and earlier, `utf8` is an alias for `utf8mb3`, the three-byte
+# variant of the unicode character set. Starting with MySQL 8, `utf8` is now an
+# alias for `utf8mb4`, the four-byte variant.
+#
+# To help ease our transition from 5.7 to 8, avoid the use of the alias
+# entirely and instead use the appropriate explicit value for the current
+# verison.
+#
+# TODO infra: once we have fully upgraded to MySQL 8+, we can remove this.
+#
+# See https://dev.mysql.com/doc/refman/8.0/en/charset-unicode-utf8.html
+module MysqlAvoidUtf8AliasInSchemaDump
+  # Reimplement ActiveRecord::SchemaDumper::format_options to replace instances
+  # of the `utf8` alias with the version-specific value it is aliasing.
+  # See https://github.com/rails/rails/blob/v6.1.7.3/activerecord/lib/active_record/schema_dumper.rb#L276-L278
+  def format_options(options)
+    options.map do |key, value|
+      if key == :charset && value == 'utf8'
+        value = current_mysql_version < 8 ? 'utf8mb3' : 'utf8mb4'
+      end
+
+      if key == :collation && value == 'utf8_unicode_ci'
+        value = current_mysql_version < 8 ? 'utf8mb3_unicode_ci' : 'utf8mb4_unicode_ci'
+      end
+
+      "#{key}: #{value.inspect}"
+    end.join(', ')
+  end
+
+  private
+
+  def current_mysql_version
+    @current_mysql_version ||= begin
+      raw_version = ActiveRecord::Base.connection.select_value('SELECT VERSION()')
+      case raw_version
+      when /^8.0.\d+/
+        8.0
+      when /^5.7.\d+/
+        5.7
+      else
+        raise "cannot parse MySQL version #{raw_version.inspect}"
+      end
+    end
+  end
+end
```

Comment:
My implementation
```
    @mysql_charset ||= begin
      version = ActiveRecord::Base.connection.select_value('SELECT VERSION()')
      major_version = version.split('.').first.to_i
      major_version >= 8 ? 'utf8mb4' : 'utf8mb3'
    rescue => e
      HoneyBadger.notify "Failed to fetch MySQL version for charset determination: #{e.message}"  
      'utf8mb3' # Default to 'utf8mb3' as a safer fallback for older versions of MySQL.
    end
```

The exception handling allows you to still get honeybadger errors but prevent any issues that may arise with a default value to the old one

Reply:
Oh, I had an underlying design decision here that I now realize is not at all clear. I'll add some comments to the code to the code explaining this, but I'll also repeat myself here:

I want this code to be explicitly temporary, and scoped only to the MySQL 5.7 -> 8.0 transition. Later versions of MySQL than 8.0 don't support `utf8mb3` in the way we need them to, and I suspect earlier versions than 5.7 also would have issues (although I haven't verified that half). My goal with this method was not to create something that could identify versions of MySQL in general, my goal was to create something that could specifically distinguish between 5.7 and 8.0, and would alarm if anything other than that showed up. If this code at any point tries to run against a MySQL database running any other version, I _want_ it to error out.

Of course, that error should probably contain more of this context and explanation than just "cannot parse". I'll make that update; thank you for calling out this lack of communication!